include/zephyr/device.h

```diff
@@ -751,10 +777,20 @@ int device_supported_foreach(const struct device *dev,
  * to an empty string, will cause NULL to be returned.
  *
  * @return pointer to device structure with the given name; NULL if
- * the device is not found or if the device with that name's
- * initialization function failed.
+ * the device is not found, if the device with that name's
+ * initialization function failed or if @kconfig{CONFIG_DEVICE_STORE_NAME}
+ * is disabled.
  */
+#ifdef CONFIG_DEVICE_STORE_NAME
 __syscall const struct device *device_get_binding(const char *name);
+#else
+static inline const struct device *device_get_binding(const char *name)
+{
+	ARG_UNUSED(name);
+
+	return NULL;
```

Comment:
This feels like a [footgun](https://en.wiktionary.org/wiki/footgun) implementation.

I believe we should not provide a device_get_binding implementation at all, at least by default, if CONFIG_DEVICE_STORE_NAME=n.

But I get why you would want this, so perhaps it should be configurable? Something like...

```
choice
	prompt "device_get_binding() behavior"
	depends on !DEVICE_STORE_NAME

config NO_DEVICE_GET_BINDING
	bool "No implementation"
	help
	  No device_get_binding() implementation will be provided.
	  Use of this function will cause build errors.

config NULL_DEVICE_GET_BINDING
	bool "NULL implementation"
	help
	  A stub device_get_binding() implementation will be provided
	  that returns NULL.

endchoice
```

with NO_DEVICE_GET_BINDING being the default behavior

Reply:
We discussed this yesterday in dev-review, agreement was to disable device_get_binding completely, so you'll get a build failure if you can't really use this option.